### ISSUE TITLE
Query Loop improvements v2

### ISSUE THREAD

**Issue by priethor** [WordPress/gutenberg#41405]
Supersedes #30910

Since the Query Loop block landed back in WordPress 5.8 and even before, numerous feedback and feature requests have been received.

The Query Loop block is such a powerful site-building tool that striking a balance between functionality and usability is critical. Therefore, ensuring optimal insertion flows and editing experience should be prioritized and considered when adding new features.

Also, leveraging block variations to consolidate similar blocks into the Query Loop will increase code reusability while improving those blocks. 

### Consolidating block variations
- [x] #32268
- [ ] #32332
- [ ] #40941
- [x] #42533

### Insertion workflows
- [x] #43427 ⚠️
- [x] #40944 (depends on #32268 and #40941)
- [ ] #37081

### Editing experience
- [x] #33202
- [ ] #37883
- [x] #30529
- [ ] #38688
- [ ] #31112
- [x] #25198
- [x] https://github.com/WordPress/gutenberg/pull/65524
- [ ] https://github.com/WordPress/gutenberg/issues/64527

### Bugs
- [ ] #42554 ⚠️

### Layouts
- [x] #34145
- [ ] #39280
- [ ] #25501
- [ ] #25502
- [ ] #25503
- [ ] #25505

### Design & Styling
- [x] #39380
- [ ] #39599

### `WP_Query` parity, functionality enhancements and other settings
- [x] #24934
- [x] #38979
- [ ] #39280
- [x] #39285
- [ ] #40903
- [x]  #41184
- [x] #40996
- [ ] #40481
- [ ] #39192
- [ ] #36785
- [x] #34201
- [ ] #42810

### Performance
- [ ] #41075

**Comment by annezazu** [comment]
Is there any chance the following can be added under editing experience https://github.com/WordPress/gutenberg/issues/25198 Based on FSE Outreach program feedback over the last few years, it's clear that more needs to be done to reimagine the various division of settings and the language used https://github.com/WordPress/gutenberg/issues/31158. This feels similar to the work done recently to improve the layout controls https://github.com/WordPress/gutenberg/pull/41893 For example, it's long been confusing for folks that to control the number of posts that show, you have to change a setting in the block toolbar rather than the sidebar where you pick which posts to display.

**Comment by padams** [comment]
It would be wonderful if we could add the attachment post type as an option (or really any custom post type or taxonomy for that matter).

**Comment by ntsekouras** [comment]
> It would be wonderful if we could add the attachment post type as an option (or really any custom post type or taxonomy for that matter).

👋  @padams - Query Loop supports custom post types and taxonomies for filtering for a while now. Were you referring to something different than filtering? 

Regarding the `attachment` option,  it is explicitly removed from the available options right now as it's a probably rare use case and would add more cognitive load to most users. Is this something you find useful and missing?

**Comment by padams** [comment]
> 👋  @padams - Query Loop supports custom post types and taxonomies for filtering for a while now. Were you referring to something different than filtering? 

Right. Just not image taxonomies...

> Regarding the `attachment` option,  it is explicitly removed from the available options right now as it's a probably rare use case and would add more cognitive load to most users. Is this something you find useful and missing?

Yes. My primary use case for WordPress is a searchable image archive. Being able to do a simple loop of the "latest images" or a loop that pulls the images for a specific custom taxonomy term would eliminate the need for a lot of plugins and custom php templates.

The killer combo IMHO would be for the gallery block to use the query block as a sort of inner block. That would give unprecedented control over galleries and keep the attachment stuff in its lane. 

Right now plug-in developers have essentially recreate the gallery block just to change the query it uses.

**Comment by NicoHood** [comment]
Please also add this one: https://github.com/WordPress/gutenberg/issues/33789 I dont know why it is closed, but it seems not yet working.

**Comment by bph** [comment]
@NicoHood could you please open a new issue with 
- a description of what you are seeing, 
- screenshot or recordings help, too, 
- the steps to reproduce and 
- information about your environment.

 That way more people can look at it again and fix whatever is happening for you.

**Comment by NicoHood** [comment]
@bph the issue I've linked describes exactly my issue. I have no idea why it was closed. It seems to be implemented, but not available for me.

**Comment by t0mtaylor** [comment]
Not sure if this can be looked at also? 

change the url of the Pagination block from /?query-8-page=2/ to /page/2/

https://wordpress.org/support/topic/how-to-change-the-url-of-the-pagination-block-from-query-8-page2-to-page-2/

**Comment by ntsekouras** [comment]
>change the url of the Pagination block from /?query-8-page=2/ to /page/2/

This is because we support multiple Query Loop blocks in the same page. So the ones that inherit the query from template do use the above format. But with multiple blocks we can have urls like this: `/page/2/?query-10-page=2`, as we preserve pagination for all the blocks.

**Comment by t0mtaylor** [comment]
> > change the url of the Pagination block from /?query-8-page=2/ to /page/2/
> 
> This is because we support multiple Query Loop blocks in the same page. So the ones that inherit the query from template do use the above format. But with multiple blocks we can have urls like this: `/page/2/?query-10-page=2`, as we preserve pagination for all the blocks.

Ah I see! Thanks @ntsekouras 

What about if there is only one query loop block on the page? could we enforce an option so it is percieved as a friendly rewrite url? `/page/2/`

Does the latest version already support `/page/2/?query-10-page=2` the friendly url stem even though it has the additional query string (as mentioned about multiple query loops per page)? If so what are the settings or functions.php snippet to do this?

You can add the rewrite url into the functions.php file, but how do you they update the url behaviour of the pagination links, next and previous, of the query loop block?

``
add_action( 'init',  function() {
  add_rewrite_rule( '^read-pg/([0-9]+)[/]?$', 'index.php?query-8-page=$matches[1]', 'top' );
} );
``

**Comment by ntsekouras** [comment]
>What about if there is only one query loop block on the page? could we enforce an option so it is percieved as a friendly rewrite url? /page/2/

I'm not really sure if we could, as the WP core handling for paging could mess with the actual query and the pagination. I don't have much experience with the url rewriting..

**Comment by mwt** [comment]
> > 👋  [@padams](https://github.com/padams) - Query Loop supports custom post types and taxonomies for filtering for a while now. Were you referring to something different than filtering?
> 
> Right. Just not image taxonomies...
> 
> > Regarding the `attachment` option,  it is explicitly removed from the available options right now as it's a probably rare use case and would add more cognitive load to most users. Is this something you find useful and missing?
> 
> Yes. My primary use case for WordPress is a searchable image archive. Being able to do a simple loop of the "latest images" or a loop that pulls the images for a specific custom taxonomy term would eliminate the need for a lot of plugins and custom php templates.
> 
> The killer combo IMHO would be for the gallery block to use the query block as a sort of inner block. That would give unprecedented control over galleries and keep the attachment stuff in its lane.
> 
> Right now plug-in developers have essentially recreate the gallery block just to change the query it uses.

This is would be super useful. If anyone has a filter to reshow the attachments, please share!

Maybe the can be an advanced option in the Gutenberg plugin?
